### PR TITLE
Ports SSpersistence, adds time-based seasons tracking system, adds rotating marine supplies

### PR DESCRIPTION
**Pull request by Walarks**
## About The Pull Request
Ports SSpersistence. Added support for automatically running events at custom time intervals, such as rotating gear in vendors.

Added some import items as gun buckets that rotate every 4 days.

## Why It's Good For The Game
More variety over time as not all items would be accessible at the same time and rotating gear can make for interesting scenarios while being done automatically.

Also adds base for implementing any other things that could progress between rounds at custom rate.

## Changelog
:cl:
code: added SSpersistence and support for real-time-based progression of stuff
/:cl:

### CODE REVIEW

**Comment by jroinc** [comment]
I like this as a concept.

File: code/__HELPERS/time.dm [hunk 1]
```diff
@@ -18,6 +18,20 @@ GLOBAL_VAR_INIT(rollovercheck_last_timeofday, 0)
...
 
 
+///Returns the first day of the given year and month in number format, from 1 (monday) - 7 (sunday).
+/proc/first_day_of_month(year, month)
+	// https://en.wikipedia.org/wiki/Zeller%27s_congruence
+	var/m = month < 3 ? month + 12 : month // month (march = 3, april = 4...february = 14)
+	var/K = year % 100 // year of century
+	var/J = round(year / 100) // zero-based century
+	// day 0-6 saturday to sunday:
+	var/h = (1 + round(13 * (m + 1) / 5) + K + round(K / 4) + round(J / 4) - 2 * J) % 7
```
Comment (BraveMole) [inline code/__HELPERS/time.dm:28]: var names

File: code/controllers/subsystem/persistence.dm [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+#define SEASONAL_GUNS "seasonal_guns"
+
+SUBSYSTEM_DEF(persistence)
+	name = "Persistence"
+	init_order = INIT_ORDER_PERSISTENCE
+	flags = SS_NO_FIRE
+
+	///Stores how long each season should last
+	var/list/seasons_durations = list(
+		SEASONAL_GUNS = 0.001 DAYS,
```
Comment (BraveMole) [inline code/controllers/subsystem/persistence.dm:15]: that's for testing i suppose

File: code/controllers/subsystem/persistence.dm [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+		)
+
+///Loads data at the start of the round
+/datum/controller/subsystem/persistence/Initialize()
+	LoadSeasonalItems()
+	return ..()
+
+///Stores data at the end of the round
+/datum/controller/subsystem/persistence/proc/CollectData()
+	return
```
Comment (BraveMole) [inline code/controllers/subsystem/persistence.dm:36]: you have plan to use that in the future?

File: code/controllers/subsystem/persistence.dm [hunk 4]
```diff
@@ -0,0 +1,105 @@
+//Season defines, universal for all seasons
+#define LAST_UPDATE "last_update" //last time the season changed
+#define CURRENT_SEASON "current_season" //current season of the section
```
Comment (BraveMole) [inline code/controllers/subsystem/persistence.dm:3]: this way of doing is weird, but i guess it's from OG persistence SS

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/controllers/subsystem/persistence.dm (hunk 2 above)
Comment (Walarks) [reply code/controllers/subsystem/persistence.dm:15]: Yes, I may remove and add this later or hopefully have some buckets I could add with this

File: code/controllers/subsystem/persistence.dm (hunk 3 above)
Comment (Walarks) [reply code/controllers/subsystem/persistence.dm:36]: Yeah, very likely that something that collects round-end info will be added

File: code/__HELPERS/time.dm (hunk 1 above)
Comment (Walarks) [reply code/__HELPERS/time.dm]: removed procs as they were unused in the end

File: code/controllers/subsystem/persistence.dm (hunk 4 above)
Comment (Walarks) [reply code/controllers/subsystem/persistence.dm:3]: discussed on discord that this is ok for now

File: code/controllers/subsystem/persistence.dm (hunk 2 above)
Comment (Walarks) [reply code/controllers/subsystem/persistence.dm]: added gun buckets that rotate every 4 days

File: code/controllers/subsystem/persistence.dm [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+	///Items that have been selected for the current round for each season
+	var/list/season_items = list()
+	///Available gun seasons
+	var/list/seasons_buckets = list(
+		SEASONAL_GUNS = list(
+		/datum/season_datum/weapons/guns/pistol1,
+		/datum/season_datum/weapons/guns/rifle1,
+		/datum/season_datum/weapons/guns/pistol2,
+		/datum/season_datum/weapons/guns/rifle2,
+		/datum/season_datum/weapons/guns/pistol3,
```
Comment (TiviPlus) [inline code/controllers/subsystem/persistence.dm:28]: Dont use numbers in typepaths

File: code/controllers/subsystem/persistence.dm [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+	var/description = "The first season."
+	///Items that the season contains
+	var/list/item_list = list()
+
+/datum/season_datum/weapons/guns/sadar_event
+	name = "SADAR bucket"
+	description = "Adds some SADARS to round-start marine gear to balance the winrates"
+	item_list = list(
+		/obj/item/weapon/gun/launcher/rocket/sadar = 10,
+		)
```
Comment (TiviPlus) [inline code/controllers/subsystem/persistence.dm:102]: uh

File: code/controllers/subsystem/persistence.dm [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+/datum/season_datum/weapons/guns/wp_event
+	name = "WP bucket"
+	description = "Adds some WP grenades to round-start marine gear to balance the winrates"
+	item_list = list(
+		/obj/item/explosive/grenade/phosphorus = 10,
+		)
```
Comment (TiviPlus) [inline code/controllers/subsystem/persistence.dm:109]: double uh

File: code/datums/gamemodes/game_mode.dm [hunk 8]
```diff
@@ -5,6 +5,8 @@
 	var/required_players = 0
 	var/maximum_players = INFINITY
 	var/squads_max_number = 4
+	///Determines whether rounds with the gamemode will be factored in when it comes to persistency
+	var/allow_persistence_save = TRUE
```
Comment (TiviPlus) [inline code/datums/gamemodes/game_mode.dm:9]: Is it like this in tg code?

File: code/game/objects/machinery/vending/vending.dm [hunk 9]
```diff
@@ -251,6 +257,10 @@
 		var/datum/vending_product/record = new(typepath = entry, product_amount = amount, product_price = prices[entry], category = category)
 		recordlist += record
 
+/obj/machinery/vending/proc/build_seasonal_tabs()
+	for(var/season in seasonal_items)
+		products[seasonal_items[season]] += SSpersistence.season_items[season]
```
Comment (TiviPlus) [inline code/game/objects/machinery/vending/vending.dm:262]: Autodoc

**Review by TiviPlus — COMMENTED** [review]
DNTM because this creates a serverside json

File: code/datums/gamemodes/game_mode.dm (hunk 8 above)
Comment (Walarks) [reply code/datums/gamemodes/game_mode.dm:9]: yes

File: code/controllers/subsystem/persistence.dm (hunk 5 above)
Comment (Walarks) [reply code/controllers/subsystem/persistence.dm]: renamed

File: code/controllers/subsystem/persistence.dm (hunk 6 above)
Comment (Walarks) [reply code/controllers/subsystem/persistence.dm]: removed

File: code/controllers/subsystem/persistence.dm (hunk 7 above)
Comment (Walarks) [reply code/controllers/subsystem/persistence.dm]: removed

File: code/game/objects/machinery/vending/vending.dm (hunk 9 above)
Comment (Walarks) [reply code/game/objects/machinery/vending/vending.dm:263]: autodocced

File: code/controllers/subsystem/persistence.dm [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+///Used to make item buckets for the seasonal items system
+/datum/season_datum
+	///Name of the  season
+	var/name = "base season"
+	///Descrpition of the season
+	var/description = "The first season."
+	///Items that the season contains
+	var/list/item_list = list()
+
+/datum/season_datum/weapons/guns/rifle_var_one
```
Comment (TiviPlus) [inline code/controllers/subsystem/persistence.dm:97]: Ok change this var to seasonal so we can avoid using a reserved word and I'll merge 
Ditto for the rest below

Comment (Walarks) [reply code/controllers/subsystem/persistence.dm:97]: changed

**Review by TiviPlus — APPROVED** [review]

**Comment by InterroLouis** [comment]
Is the extra smartmachinegun drum in their box intentional?

**Pull request merged**